Remove odo push command

feloy opened the pull request:
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

- Remove odo push command
- Move integration tests from 'odo push' to 'odo dev'
- Some remaining tests commented in `cmd_devfile_push_test.go` need to be ported later to `odo dev` when the behaviour is supported

**Which issue(s) this PR fixes:**

Fixes #5560

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

Blocked by #5569 for replacing some `odo push` with `odo dev`  tests (the tests that are checking the behaviour of `odo dev/push` with specific information in the devfile (endpoints, volumes, etc)

rm3l commented on tests/integration/devfile/cmd_devfile_push_test.go:
```diff
@@ -1,5 +1,6 @@
 package devfile
```
I guess this file is no longer necessary. Shouldn't we remove it then?
Also, how about also removing the corresponding `test-cmd-devfile-push` Makefile target?

feloy replied:
I would like to rework some of these tests with `odo dev`

rm3l replied:
Ok, I see. Then maybe a simple comment in the file could help keep track of this?

feloy replied:
I'll make the changes in this PR, when #5569 is merged

rm3l reviewed (changes requested):
Looks like there is still an occurrence of `odo push` in the help output:

```
❯ odo help | grep -A3 -B3 push
  # Creating and deploying a Node.js project
  git clone https://github.com/odo-devfiles/nodejs-ex && cd nodejs-ex
  odo create nodejs
  odo push
  
  # Accessing your Node.js component
  odo url create
```

Should we remove this from the help message as well?

feloy commented:
> Looks like there is still an occurrence of `odo push` in the help output:
> 
> ```
> ❯ odo help | grep -A3 -B3 push
>   # Creating and deploying a Node.js project
>   git clone https://github.com/odo-devfiles/nodejs-ex && cd nodejs-ex
>   odo create nodejs
>   odo push
>   
>   # Accessing your Node.js component
>   odo url create
> ```
> 
> Should we remove this from the help message as well?

Most of the help output is to be rewritten. I would prefer it is part of another PR about doc

rm3l commented:
> > Looks like there is still an occurrence of `odo push` in the help output:
> > ```
> > ❯ odo help | grep -A3 -B3 push
> >   # Creating and deploying a Node.js project
> >   git clone https://github.com/odo-devfiles/nodejs-ex && cd nodejs-ex
> >   odo create nodejs
> >   odo push
> >   
> >   # Accessing your Node.js component
> >   odo url create
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should we remove this from the help message as well?
> 
> Most of the help output is to be rewritten. I would prefer it is part of another PR about doc

Ok, makes sense. I spotted this after reading the acceptance criteria of #5570 , which stated that "_all help messages and error messages need to be reviewed and updated to make sure that it doesn't contain a reference to commands that no longer exist_"

feloy commented:
> > > Looks like there is still an occurrence of `odo push` in the help output:
> > > ```
> > > ❯ odo help | grep -A3 -B3 push
> > >   # Creating and deploying a Node.js project
> > >   git clone https://github.com/odo-devfiles/nodejs-ex && cd nodejs-ex
> > >   odo create nodejs
> > >   odo push
> > >   
> > >   # Accessing your Node.js component
> > >   odo url create
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Should we remove this from the help message as well?
> > 
> > 
> > Most of the help output is to be rewritten. I would prefer it is part of another PR about doc
> 
> Ok, makes sense. I spotted this after reading the acceptance criteria of #5560 , which stated that "_all help messages and error messages need to be reviewed and updated to make sure that it doesn't contain a reference to commands that no longer exist_".

You are right, I have made changes to try and remove all references to `odo push`

dharmit commented on pkg/component/component.go:
```diff
@@ -202,28 +202,6 @@ func GetComponentTypeFromDevfileMetadata(metadata devfile.DevfileMetadata) strin
 	return componentType
 }
 
-// GetProjectTypeFromDevfileMetadata returns component type from the devfile metadata
-func GetProjectTypeFromDevfileMetadata(metadata devfile.DevfileMetadata) string {
-	var projectType string
-	if metadata.ProjectType != "" {
-		projectType = metadata.ProjectType
-	} else {
-		projectType = NotAvailable
-	}
-	return projectType
-}
-
-// GetLanguageFromDevfileMetadata returns component type from the devfile metadata
-func GetLanguageFromDevfileMetadata(metadata devfile.DevfileMetadata) string {
-	var language string
-	if metadata.Language != "" {
-		language = metadata.Language
-	} else {
-		language = NotAvailable
-	}
-	return language
-}
```
IIRC, these functions were used for `odo push`'s telemetry part. We are going to need this again because we will need telemetry for `odo dev` as well. ATM, `odo dev` doesn't have telemetry (which occurred to me when I saw these functions being removed.)

I don't know if we should keep these functions for above stated future use or not, I said this just as a note.

feloy replied:
Right, it seems safer to keep them

dharmit commented on pkg/odo/cli/component/common_push.go:
```diff
@@ -5,122 +5,12 @@ import (
-	return componentNamespace, nil
-}
-
 // GatherName parses the Devfile and retrieves an appropriate name in two ways.
 // 1. If metadata.name exists, we use it
 // 2. If metadata.name does NOT exist, we use the folder name where the devfile.yaml is located
```
Thinking out loud:

This file now has only one function `GatherName`. Does this function belong more in `libdevfile` package? Secondly, and more importantly, I wonder if any devfiles in the default registry don't have `metadata.name`. This function was useful in the past when all devfiles didn't have this field. But now, IMO, it's safe to replace calls to this function with just `devfileObj.Data.GetMetadata().Name`.

I'm not suggesting to address this in this PR, but we could open a separate `good-first-issue` if what I am saying makes sense.

feloy replied:
I moved it to the `pkg/component` package in the meantime.

I prefer not to move it to libdevfile, as it contains some odo logic IMO.

dharmit replied:
Indeed, indeed. The first part of just getting name is straightforward, doesn't need a separate function; but the second part is odo specific. Although what I was thinking is maybe we could get rid of it and see because devfiles these days do have `.metadaata.name` defined. And this could be a separate issue.

dharmit commented on tests/helper/helper_dev.go:
```diff
@@ -105,10 +127,23 @@ func (o DevSession) Stop() {
 
+func (o DevSession) WaitSync() ([]byte, []byte) {
+	WaitForOutputToContain("Pushing files...", 180, 10, o.session)
+	WaitForOutputToContain("Waiting for something to change", 180, 10, o.session)
```
This message will change once #5526 gets merged. Just FYI.

feloy replied:
As this message is not only informational but used for tests to work correctly, I think it would be nice to have a constant defined in the source. I'll add it once #5526 is merged

dharmit replied:
Sounds good. 👍🏾

dharmit commented on tests/helper/helper_dev.go:
```diff
@@ -105,10 +127,23 @@ func (o DevSession) Stop() {
+	return out, err
+}
+
+func (o DevSession) WaitForOutputToContain(substring string, timeoutInSeconds int, intervalInSeconds int) {
```
What's the purpose of this function? I don't see this getting used anywhere in the tests:
```sh
$ grep -r "WaitForOutputToContain" tests/integration
tests/integration/devfile/cmd_dev_test.go:				helper.WaitForOutputToContain("Error occurred on Push", 180, 10, session)
tests/integration/devfile/cmd_dev_test.go:			helper.WaitForOutputToContain("Waiting for something to change", 180, 10, session)
tests/integration/devfile/cmd_dev_test.go:			helper.WaitForOutputToContain("server.js", 180, 10, session)
tests/integration/devfile/cmd_dev_test.go:				helper.WaitForOutputToContain("Waiting for something to change", 180, 10, session)
tests/integration/devfile/cmd_dev_test.go:					helper.WaitForOutputToContain("Waiting for something to change", 180, 10, session)
tests/integration/devfile/cmd_delete_test.go:				helper.WaitForOutputToContain("Press Ctrl+c to exit", 180, 10, session)
tests/integration/devfile/cmd_delete_test.go:			helper.WaitForOutputToContain("Press Ctrl+c to exit", 180, 10, session)
```

feloy replied:
Not used anymore, I remove it

dharmit reviewed (commented):
Mostly comments/questions. Not really change requests.

dharmit commented:
Looks like legit errors in unit tests:
```
pkg/odo/cli/cli.go:16:2: dev redeclared as imported package name
	previous declaration at pkg/odo/cli/cli.go:10:2 
```

dharmit commented on tests/helper/helper_dev.go:
```diff
@@ -77,22 +88,33 @@ import (
 }
 
 // StartDevMode starts a dev session with `odo dev`
-func StartDevMode() DevSession {
+func StartDevMode() (DevSession, []byte, []byte) {
```
Can you update the comment to include info about the two `[]byte`s we return here?

feloy commented:
Thanks @dharmit for your review

dharmit commented:
@feloy do you think we should port some commented tests from https://github.com/redhat-developer/odo/blob/8a3ae3abfb752ac0afdf00ab4eea2265448e0e21/tests/integration/devfile/cmd_devfile_push_test.go to `odo dev` tests? A few that I think could be ported (and hence removed from the commented section) are:
* `When("setting git config and running odo push",`
* `When("the Deployment's Replica count (pods) is scaled to 0",`
* `When("doing odo push and run command is not marked as hotReloadCapable"`
* `Context("devfile is modified",`
* `When("creating nodejs component, doing odo push and run command has dev.odo.push.path attribute",`
* `When("project with with 'default' name is used",` 

I haven't looked at all the tests within the above-mentioned specs, and I have also not looked completely at the tests you have added to `odo dev`. So, if you have already added some of these to `odo dev`, I would suggest removing them from comment (knowing you, I think you would have already done if that was the case.) If you think they shouldn't be ported to `odo dev` now, when do you suggest we port it?

feloy commented:
Here are the tests I removed or ported to the `cmd_dev_test.go` file

```
var _ = Describe("odo devfile push command tests", func() {

	When("creating a nodejs component", func() {
		When("setting git config and running odo push", func() {
			It("checks that odo push works with a devfile", func() {
				// removed
			})

			It("check annotations from the deployment after odo push", func() {
				// ported to "should create vcs-uri annotation for the deployment when running odo dev"
			})

			When("updating a variable into devfile", func() {
				It("should run odo push successfully", func() {
					// removed
				})

			})
		})

		When("odo push is executed with json output", func() {
			// removed
			It("should display push output in JSON format", func() {
			})
			When("update devfile and push again", func() {

				It("should display push updated output in JSON format", func() {
				})
			})
		})

		When("running odo push outside the context directory", func() {
			// removed
			It("should push correct component based on --context flag", func() {
			})
		})

		When("Devfile 2.1.0 is used", func() {

			When("doing odo push", func() {
				It("should pass", func() {
					// removed
				})
				It("should check if the env variable has a correct value", func() {
					// ported to "should check if the env variable has a correct value"
				})
			})
		})
		When("doing odo push", func() {
			When("doing odo push with -f flag", func() {
				It("should build even when no changes are detected", func() {
					// removed
				})
			})

			When("the pod is deleted and replaced", func() {
				It("should execute run command on odo push", func() {
					// removed (stopping odo dev will delete the deployment)
				})
			})

			When("the Deployment's Replica count (pods) is scaled to 0", func() {
				It("odo push should successfully recreate the pod", func() {
					// removed (stopping odo dev will delete the deployment)
				})
			})
		})

		When("creating local files and dir and doing odo push", func() {
			// ported to "creating local files and dir and running odo dev"
			It("should correctly propagate changes to the container", func() {
				// ported
			})
			When("deleting local files and dir and doing odo push again", func() {
				It("should not list deleted dir and file in container", func() {
					// ported
				})
			})
		})

		When("devfile has sourcemappings and doing odo push", func() {
			It("should sync files to the correct location", func() {
				// ported
			})
		})

		When("project and clonePath is present in devfile and doing odo push", func() {
			It("should sync to the correct dir in container", func() {
				// ported
			})
		})

		When("devfile project field is present and doing odo push", func() {
			It("should sync to the correct dir in container", func() {
				// ported
			})
		})

		When("multiple project is present", func() {
			It("should sync to the correct dir in container", func() {
				// ported
			})
		})

		When("no project is present", func() {
			It("should sync to the correct dir in container", func() {
				// ported
			})
		})

		When("doing odo push with devfile contain volume", func() {
			It("should create pvc and reuse if it shares the same devfile volume name", func() {
			// ported
		})
			It("check the volume name and mount paths for the containers", func() {
			// ported
		})

		})

		When("doing odo push with devfile containing volume-component", func() {
			It("should successfully use the volume components in container components", func() {
			// ported
		})
		})

		When("doing odo push and run command is not marked as hotReloadCapable", func() {
			It("should restart the application", func() {
				// ?
			})
		})

		When("doing odo push and run command is marked as hotReloadCapable:true", func() {
			It("should not restart the application", func() {
				// ?
			})

			When("doing odo push --debug ", func() {
				It("should restart the application regardless of hotReloadCapable value", func() {
					// removed
				})
			})
		})

		When("doing odo push and devfile with composite command", func() {
			It("should execute all commands in composite commmand", func() {
				// ported
			})
		})
		When("doing odo push and composite command is marked as paralell:true ", func() {
			It("should execute all commands in composite commmand", func() {
				// ported
			})
		})

		When("doing odo push and composite command are nested", func() {
			It("should execute all commands in composite commmand", func() {
				// ported
			})
		})
		When("doing odo push and composite command is used as a run command", func() {
			It("should throw a validation error for composite run commands", func() {
				// ported
			})
		})

		When("events are defined", func() {

			It("should correctly execute PreStart commands", func() {
				// ported
			})

			It("should correctly execute PostStart commands", func() {
				// removed
			})
		})

		When("doing odo push and using wrong custom commands (specified by flags)", func() {
			It("should error out", func() {
				// removed
			})

		})

		When("command has no group kind and doing odo push", func() {
			It("should execute commands with flags", func() {
				// removed
			})
		})

		When("doing odo push and run command throws an error", func() {
			It("should wait and error out with some log", func() {
				// ported
			})
		})

		When("commands specify have env variables", func() {
			When("doing odo push and sigle env var is set", func() {
				It("should be able to exec command", func() {
					// ported
				})
			})
			When("doing odo push and multiple env variables are set", func() {
				It("should be able to exec command", func() {
					// ported
				})
			})
			When("doing odo push and there is a env variable with spaces", func() {
				It("should be able to exec command", func() {
					// ported
				})
			})
		})
	})

	Context("pushing devfile without an .odo folder", func() {
		It("should error out on odo push and passing invalid devfile", func() {
			//removed
		})
	})

	When("Create and push java-springboot component", func() {
		It("should execute default build and run commands correctly", func() {
			// ported
		})
	})

	Context("devfile is modified", func() {
		// Tests https://github.com/redhat-developer/odo/issues/3838
		
		Context("odo push -f is executed", func() {
			ensureFilesSyncedTest(commonVar.Project, true)
			// removed
		})
		Context("odo push (without -f) is executed", func() {
			ensureFilesSyncedTest(commonVar.Project, false)
			// ported
		})

		When("node-js application is created and pushed with devfile schema 2.2.0", func() {

		
			It("should check cpuLimit,cpuRequests,memoryRequests", func() {
				ensureResource(output, "1", "200m", "512Mi")
				// ported
			})

			When("Update the devfile.yaml, do odo push", func() {

		
				It("should check cpuLimit,cpuRequests,memoryRequests", func() {
					ensureResource(output, "700m", "250m", "550Mi")
					// ported
				})
			})
		})
	})

	When("creating nodejs component, doing odo push and run command has dev.odo.push.path attribute", func() {
		It("should push only the mentioned files at the appropriate remote destination", func() {
			// ported
		})
	})

	Context("using OpenShift cluster", func() {
		When("project with with 'default' name is used", func() {
			It("should throw an error", func() {
				// TODO 
			})
		})
	})

	Context("using Kubernetes cluster", func() {
		When("project with with 'default' name is used", func() {

			It("should push successfully", func() {
				// ported
			})
		})
	})

})
```

feloy commented:
@valaparthvi 
> Is waiting for Ctrl+C to appear correct ? I have used it in the past because it seems convenient.

For the code reader, I find more understandable to wait for "Waiting for something to change" instead of "Press Ctrl+c to exit", as the first one is indicating that odo is in a stable state, and waiting for some changes in the files, when the latter does not indicate this state (and could be displayed elsewhere in the output)

rm3l reviewed (commented):
Out of curiosity, what is the meaning of `MR-CL-CR` in the name of this file: `tests/examples/source/devfiles/nodejs/devfile-with-MR-CL-CR-modified.yaml`?

I guess it is about memory and cpu limits and requests, right?
I had to look at the test where it is being used to guess that, so maybe the file name could reflect that.. :)

valaparthvi commented:
/lgtm

valaparthvi commented:
/approve
All the tests pass.
